agents/agents-core/src/commonMain/kotlin/ai/koog/agents/core/agent/ActAIAgent.kt

```diff
@@ -0,0 +1,213 @@
+@file:OptIn(ExperimentalUuidApi::class)
+
+package ai.koog.agents.core.agent
+
+import ai.koog.agents.core.agent.ActAIAgent.FeatureContext
+import ai.koog.agents.core.agent.config.AIAgentConfig
+import ai.koog.agents.core.agent.config.AIAgentConfigBase
+import ai.koog.agents.core.agent.context.AIAgentLLMContext
+import ai.koog.agents.core.agent.context.element.AgentRunInfoContextElement
+import ai.koog.agents.core.agent.entity.AIAgentStateManager
+import ai.koog.agents.core.agent.entity.AIAgentStorage
+import ai.koog.agents.core.environment.GenericAgentEnvironment
+import ai.koog.agents.core.feature.AIAgentFeature
+import ai.koog.agents.core.feature.AIAgentNonGraphFeature
+import ai.koog.agents.core.feature.AIAgentNonGraphPipeline
+import ai.koog.agents.core.feature.PromptExecutorProxy
+import ai.koog.agents.core.feature.config.FeatureConfig
+import ai.koog.agents.core.tools.ToolRegistry
+import ai.koog.prompt.executor.clients.openai.OpenAIModels
+import ai.koog.prompt.executor.model.PromptExecutor
+import ai.koog.prompt.llm.LLModel
+import io.github.oshai.kotlinlogging.KotlinLogging
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import kotlinx.datetime.Clock
+import kotlin.uuid.ExperimentalUuidApi
+import kotlin.uuid.Uuid
+
+/**
+ * Represents the core AI agent for processing input and generating output using
+ * a defined configuration, toolset, and prompt execution pipeline.
+ *
+ * @param Input The type of input data expected by the agent.
+ * @param Output The type of output data produced by the agent.
+ * @param id The unique identifier for the agent instance.
+ * @param promptExecutor The executor responsible for processing prompts and interacting with language models.
+ * @param agentConfig The configuration for the agent, including the prompt structure and execution parameters.
+ * @param toolRegistry The registry of tools available for the agent. Defaults to an empty registry if not specified.
+ */
+@ExperimentalUuidApi
+public class ActAIAgent<Input, Output>(
```

Comment:
Also, why is it an `ActAIAgent` ? Again, if it's a non-graph generic agent that can do anything it wants inside the strategy, it's not Act or ReAct -- so we probably need some more general name here

Reply:
Any suggestions on the name? We didn't have better name then ActAgent (NonGraphAgent sounds horrible for me)

Reply:
My take from highest to lowest preference ProceduralAgent, ProgramaticAgent, CodeBasedAgent, FunctionalAgent.

Reply:
Yeah, I had put FunctionalAgent at the end of my list because I was afraid it might imply to some users that they are "pure", i.e. without side-effect, which they obviously are not. But functions in Kotlin are not pure, so yeah it's not the worst name.